File: WallTeleporters/src/unwrittenfun/minecraft/wallteleporters/blocks/BlockWallTeleporter.java
package unwrittenfun.minecraft.wallteleporters.blocks;

import cpw.mods.fml.common.network.FMLNetworkHandler;
import net.minecraft.block.Block;
import net.minecraft.block.BlockContainer;
import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.texture.IconRegister;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.Icon;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import unwrittenfun.minecraft.wallteleporters.WallTeleporters;
import unwrittenfun.minecraft.wallteleporters.blocks.tileentities.TileEntityWallTeleporter;
import unwrittenfun.minecraft.wallteleporters.client.gui.creativetabs.WTCreativeTabs;
import unwrittenfun.minecraft.wallteleporters.info.BlockInfo;
import unwrittenfun.minecraft.wallteleporters.info.ModInfo;

/**
 * Mod: Wall Teleporters
 * Author: UnwrittenFun
 * License: Minecraft Mod Public License (Version 1.0.1)
 */
public class BlockWallTeleporter extends BlockContainer {
    public BlockWallTeleporter(int id) {
        super(id, Material.iron);

        setUnlocalizedName(BlockInfo.WT_UNLOCAL_NAME);
        setHardness(2F);
        setCreativeTab(WTCreativeTabs.wtTab);
    }

    @Override
    public void registerIcons(IconRegister register) {
        blockIcon = register.registerIcon(ModInfo.TEXTURE_LOCATION + ":" + BlockInfo.WT_ICON_NAME);
    }

    @Override
    public Icon getBlockTexture(IBlockAccess world, int x, int y, int z, int side) {
        TileEntity tileEntity = world.getBlockTileEntity(x, y, z);

        if (tileEntity instanceof TileEntityWallTeleporter) {
            TileEntityWallTeleporter teleporter = ((TileEntityWallTeleporter) tileEntity);

            if (teleporter.mask[0] != 0) {
                Block maskBlock = Block.blocksList[teleporter.mask[0]];
                return maskBlock.getIcon(side, teleporter.mask[1]);
            }
        }

        return blockIcon;
    }

    @Override
    public TileEntity createNewTileEntity(World world) {
        return new TileEntityWallTeleporter();
    }

    @Override
    public AxisAlignedBB getCollisionBoundingBoxFromPool(World world, int x, int y, int z) {
        return AxisAlignedBB.getAABBPool().getAABB(0, 0, 0, 0, 0, 0);
    }

    @Override
    public boolean isOpaqueCube() {
        return false;
    }

    @Override
    public boolean renderAsNormalBlock() {
        return false;
    }

    @Override
    public void onEntityCollidedWithBlock(World world, int x, int y, int z, Entity entity) {
        if (!world.isRemote) {
            if (entity instanceof EntityPlayerMP) {
                EntityPlayerMP player = (EntityPlayerMP) entity;
                TileEntity tileEntity = world.getBlockTileEntity(x, y, z);

                if (tileEntity instanceof TileEntityWallTeleporter) {
                    TileEntityWallTeleporter teleporter = (TileEntityWallTeleporter) tileEntity;
                    teleporter.teleportPlayer(player);
                }
            }
        }
    }

    @Override
    public void onBlockAdded(World world, int x, int y, int z) {
        if (!world.isRemote) {
            TileEntity tileEntity = world.getBlockTileEntity(x, y, z);

            if (tileEntity instanceof TileEntityWallTeleporter) {
                TileEntityWallTeleporter teleporter = (TileEntityWallTeleporter) tileEntity;
                teleporter.notifyNeighboursOfConnectionChanged();
            }
        }
    }

    @Override
    public boolean onBlockActivated(World world, int x, int y, int z, EntityPlayer player,
                                    int side, float hitX, float hitY, float hitZ) {
        boolean result = true;
        TileEntity tileEntity = world.getBlockTileEntity(x, y, z);

        if (tileEntity instanceof TileEntityWallTeleporter) {
            TileEntityWallTeleporter teleporter = ((TileEntityWallTeleporter) tileEntity);

            ItemStack held = player.inventory.getCurrentItem();
            if (!teleporter.multiblock.isLocked() && held != null && held.getItem() instanceof ItemBlock) {
                if (((ItemBlock) held.getItem()).getBlockID() == WTBlocks.wallTeleporter.blockID &&
                    teleporter.mask[0] == 0) {
                    result = false;
                } else if (!world.isRemote) {
                    teleporter.setMask(((ItemBlock) held.getItem()).getBlockID(), held.getItemDamage());
                }
            } else {
                FMLNetworkHandler.openGui(player, WallTeleporters.instance, 0, world, x, y, z);
            }
        }

        return result;
    }

    @Override
    public void breakBlock(World world, int x, int y, int z, int id, int meta) {
        TileEntity tileEntity = world.getBlockTileEntity(x, y, z);
        if (tileEntity instanceof TileEntityWallTeleporter) {
            IInventory inventory = (IInventory) tileEntity;

            for (int i = 0; i < inventory.getSizeInventory(); i++) {
                ItemStack stack = inventory.getStackInSlotOnClosing(i);

                if (stack != null) {
                    float spawnX = x + world.rand.nextFloat();
                    float spawnY = y + world.rand.nextFloat();
                    float spawnZ = z + world.rand.nextFloat();

                    EntityItem droppedItem = new EntityItem(world, spawnX, spawnY, spawnZ, stack);

                    float mult = 0.05F;

                    droppedItem.motionX = (-0.5F + world.rand.nextFloat()) * mult;
                    droppedItem.motionY = (4 + world.rand.nextFloat()) * mult;
                    droppedItem.motionZ = (-0.5F + world.rand.nextFloat()) * mult;

                    world.spawnEntityInWorld(droppedItem);
                }
            }
        }

        super.breakBlock(world, x, y, z, id, meta);
    }
}

File: scala/unwrittenfun/minecraft/unwrittenblocks/blocks/BlockWallTeleporter.scala
package unwrittenfun.minecraft.unwrittenblocks.blocks

import net.minecraft.block.{Block, BlockContainer}
import net.minecraft.block.material.Material
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.client.renderer.texture.IconRegister
import unwrittenfun.minecraft.unwrittenblocks.{UnwrittenBlocks, TEXTURE_LOCATION}
import net.minecraft.util.{AxisAlignedBB, Icon}
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.{World, IBlockAccess}
import unwrittenfun.minecraft.unwrittenblocks.blocks.tileentities.TileEntityWallTeleporter
import net.minecraft.entity.player.{EntityPlayer, EntityPlayerMP}
import net.minecraft.entity.Entity
import net.minecraft.item.{ItemBlock, ItemStack}
import cpw.mods.fml.common.network.FMLNetworkHandler
import net.minecraft.inventory.IInventory
import net.minecraft.entity.item.EntityItem
import unwrittenfun.minecraft.unwrittenblocks.client.gui.creativetabs

/**
 * Mod: UnwrittenBlocks
 * Author: UnwrittenFun
 * License: Minecraft Mod Public License (Version 1.0.1)
 */
class BlockWallTeleporter(id: Int, key: String) extends BlockContainer(id, Material.iron) {
  setUnlocalizedName(key)
  setHardness(2f)
  setCreativeTab(creativetabs.ubTab)

  override def registerIcons(register: IconRegister) {
    blockIcon = register.registerIcon(TEXTURE_LOCATION + ":" + key)
  }

  override def getBlockTexture(world: IBlockAccess, x: Int, y: Int, z: Int, side: Int): Icon = {
    var rtn: Icon = blockIcon
    world.getBlockTileEntity(x, y, z) match {
      case teleporter: TileEntityWallTeleporter =>
        if (teleporter.mask(0) != 0) {
          val maskBlock = Block.blocksList(teleporter mask 0)
          rtn = maskBlock.getIcon(side, teleporter mask 1)
        }
      case _ =>
    }
    rtn
  }

  override def createNewTileEntity(world: World): TileEntity = new TileEntityWallTeleporter

  override def getCollisionBoundingBoxFromPool(world: World, x: Int, y: Int, z: Int): AxisAlignedBB = AxisAlignedBB.getAABBPool.getAABB(0, 0, 0, 0, 0, 0)

  override def isOpaqueCube: Boolean = false

  override def renderAsNormalBlock(): Boolean = false

  override def onEntityCollidedWithBlock(world: World, x: Int, y: Int, z: Int, entity: Entity) {
    if (!world.isRemote) {
      entity match {
        case player: EntityPlayerMP =>
          world.getBlockTileEntity(x, y, z) match {
            case teleporter: TileEntityWallTeleporter => if (teleporter.multiblock.getTrips > 0) teleporter.teleportPlayer(player)
            case _ =>
          }
        case _ =>
      }
    }
  }

  override def onBlockAdded(world: World, x: Int, y: Int, z: Int) {
    if (!world.isRemote) {
      world.getBlockTileEntity(x, y, z) match {
        case teleporter: TileEntityWallTeleporter => teleporter.notifyNeighboursOfConnectionChanged()
        case _ =>
      }
    }
  }

  override def onBlockActivated(world: World, x: Int, y: Int, z: Int, player: EntityPlayer, side: Int, hitX: Float, hitY: Float, hitZ: Float): Boolean = {
    var result: Boolean = true
    world.getBlockTileEntity(x, y, z) match {
      case teleporter: TileEntityWallTeleporter =>
        val held: ItemStack = player.inventory.getCurrentItem
        if (!teleporter.multiblock.locked && held != null && held.getItem.isInstanceOf[ItemBlock]) {
          if (held.getItem.asInstanceOf[ItemBlock].getBlockID == UBBlocks.wallTeleporter.blockID && teleporter.mask(0) == 0) {
            result = false
          }
          else if (!world.isRemote) {
            teleporter.setMask(held.getItem.asInstanceOf[ItemBlock].getBlockID, held.getItemDamage)
          }
        }
        else {
          FMLNetworkHandler.openGui(player, UnwrittenBlocks, 0, world, x, y, z)
        }
      case _ =>
    }
    result
  }

  override def breakBlock(world: World, x: Int, y: Int, z: Int, id: Int, meta: Int) {
    UBBlocks.dropItemsFromInventory(world, x, y, z)
    super.breakBlock(world, x, y, z, id, meta)
  }
}
